x/stakeibc/keeper/icacallbacks_undelegate.go

```diff
@@ -81,36 +79,70 @@ func (k Keeper) UndelegateCallback(ctx sdk.Context, packet channeltypes.Packet,
 	k.Logger(ctx).Info(utils.LogICACallbackStatusWithHostZone(chainId, ICACallbackID_Undelegate,
 		icacallbackstypes.AckResponseStatus_SUCCESS, packet))
 
-	// Update delegation balances
-	err := k.UpdateDelegationBalances(ctx, hostZone, undelegateCallback)
+	// Calculate the stTokens that should be burned from the batch and get the latest
+	// completion time from the ack response
+	nativeTokensUnbonded := k.CalculateTotalUnbondedInBatch(undelegateCallback.SplitUndelegations)
+	unbondingTime, err := k.GetLatestUnbondingCompletionTime(ctx, ackResponse.MsgResponses)
 	if err != nil {
 		k.Logger(ctx).Error(fmt.Sprintf("UndelegateCallback | %s", err.Error()))
 		return err
 	}
 
-	// Get the latest transaction completion time (to determine the unbonding time)
-	latestCompletionTime, err := k.GetLatestCompletionTime(ctx, ackResponse.MsgResponses)
+	// Update delegation balances on the validators and host zone
+	err = k.UpdateDelegationBalances(ctx, hostZone, undelegateCallback)
 	if err != nil {
 		k.Logger(ctx).Error(fmt.Sprintf("UndelegateCallback | %s", err.Error()))
 		return err
 	}
 
-	// Burn the stTokens
-	stTokenBurnAmount, err := k.UpdateHostZoneUnbondings(ctx, *latestCompletionTime, chainId, undelegateCallback)
+	// Update the accounting on the host zone unbondings
+	stTokensToBurn, err := k.UpdateHostZoneUnbondingsAfterUndelegation(
+		ctx,
+		chainId,
+		undelegateCallback.EpochUnbondingRecordIds,
+		nativeTokensUnbonded,
+		unbondingTime,
+	)
 	if err != nil {
 		k.Logger(ctx).Error(fmt.Sprintf("UndelegateCallback | %s", err.Error()))
 		return err
 	}
-	err = k.BurnTokens(ctx, hostZone, stTokenBurnAmount)
-	if err != nil {
+
+	// Burn the stTokens from the batch
+	if err := k.BurnStTokensAfterUndelegation(ctx, hostZone, stTokensToBurn); err != nil {
 		k.Logger(ctx).Error(fmt.Sprintf("UndelegateCallback | %s", err.Error()))
 		return err
 	}
 
-	// Upon success, add host zone unbondings to the exit transfer queue
-	err = k.RecordsKeeper.SetHostZoneUnbondingStatus(ctx, chainId, undelegateCallback.EpochUnbondingRecordIds, recordstypes.HostZoneUnbonding_EXIT_TRANSFER_QUEUE)
-	if err != nil {
-		return err
+	return nil
+}
+
+// Regardless of failure/success/timeout, indicate that this ICA has completed on each validator
+// on the host zone, and on the epoch unbonding record
+func (k Keeper) MarkUndelegationAckReceived(ctx sdk.Context, hostZone types.HostZone, undelegateCallback types.UndelegateCallback) error {
+	// Indicate that this ICA has completed on each validator
+	for _, splitDelegation := range undelegateCallback.SplitUndelegations {
+		if err := k.DecrementValidatorDelegationChangesInProgress(&hostZone, splitDelegation.Validator); err != nil {
+			return err
```

Comment:
any error will get logged upstream in the icacallback module!  Lemme know if you think I should add some more context to the message though

In general the pattern we have in stakeibc where we log and then return the error is kind of an antipattern from the early days where we were being extra cautious. Everything should be setup now so that you can just return an error and then it will get logged by the calling function!

Reply:
Ah gotcha, this makes sense! I'm onboard with not adding additional print statements then